### PR TITLE
feat: Discord authentication

### PR DESCRIPTION
**Pull request by karelianpie**
## Motivation and Context

Create a connection between Discord and Coordinape app

## Description

* Create bot schema;
* Create a simple page for linking.

## Test and Deployment Plan

In order to test this, you'll need to create your own discord bot https://discord.com/developers/applications

Once you have a bot, add the following environment variables to `.env`

```
DISCORD_BOT_CLIENT_ID=
DISCORD_BOT_CLIENT_SECRET=
DISCORD_BOT_REDIRECT_URI=http://localhost:3000/discord/link
```

You can find the client id and secret in the `OAuth2 > General` section of your [application](https://discord.com/developers/applications)

Start the app with `yarn start`

Enable the feature flag 'discord'

Go back to the discord application page, navigate to `OAuth2 > URL Generator` and in "Redirects" add the following URLs:

http://localhost:3000/discord/link

Then, go to "URL Generator", select the "Identity" scope, select the url http://localhost:3000/discord/link and copy the generated url to your browser.

You should be redirected to https://discord.com/oauth2/authorize where you can authorize the bot/application.

Once you're authorized you should be redirected to coordinape where you'll need to click the "Link" button, once clicked your snowflake id will be stored in coordinape's database.

## Screenshots (if appropriate):

https://user-images.githubusercontent.com/78794805/202471324-183bf5de-8cfb-4bb1-94a5-72b448c82f49.mov

## Reviewers

@topocount 

## Related Issue

Closes https://github.com/coordinape/discord-bot/issues/31

### CODE REVIEW

File: src/routes/routes.tsx
Comment (topocount) [inline src/routes/routes.tsx:87]: Is this just for prototyping? I think this flow makes sense, but we're going to receive the token via [discord's OAuth flow](https://discordjs.guide/oauth2/#authorization-code-grant-flow), right?

**Review by topocount — COMMENTED** [review]
@karelianpie do you need any advice or assistance here? I'm happy to help get tests passing, for one.

I think the dataflow makes sense, but it's not a proper OAuth implementation yet so I'll hold off on merging, unless we cut the frontend pieces for now. I'm pretty confident the /co link command will just kick off the web-based flow that'll require authentication. I'll link you to the designs for the auth flow when I have the finished set of designs.

File: src/routes/routes.tsx
Comment (karelianpie) [reply src/routes/routes.tsx:88]: Correct, implemented an MVP of the discord's authorization code grant flow with https://github.com/coordinape/coordinape/pull/1625/commits/0ed646842159f67e192d622b816c5fdbcbfb837c

We will get the following JSON as a response:
```json
{
	"access_token": "an access token",
	"token_type": "Bearer",
	"expires_in": 604800,
	"refresh_token": "a refresh token",
	"scope": "identify"
}
```
Was the idea to add that data to the discord's `users` table? We will need to keep the refresh token in hand, so when the original access token expires we will need to request a new one [like so](https://discord.com/developers/docs/topics/oauth2#authorization-code-grant-refresh-token-exchange-example).

**Comment by karelianpie** [comment]
> @karelianpie do you need any advice or assistance here? I'm happy to help get tests passing, for one.

I think the tests failing is not related to this PR

> I think the dataflow makes sense, but it's not a proper OAuth implementation yet so I'll hold off on merging, unless we cut the frontend pieces for now. The /co link command will just kick off the web-based flow, I believe. I'll link you to the designs for the auth flow when I have the finished set of designs.

Sounds good, this is just a draft for the time being. Let's get the OAuth implementation ready before merge, already pushed https://github.com/coordinape/coordinape/pull/1625/commits/0ed646842159f67e192d622b816c5fdbcbfb837c as a starting point

File: hasura/migrations/default/1654202757203_alter_table_public_circle_api_keys_add_column_read_discord/down.sql
Comment (topocount) [inline hasura/migrations/default/1654202757203_alter_table_public_circle_api_keys_add_column_read_discord/down.sql:4]: let's delete this boilerplate.

File: src/routes/routes.tsx
Comment (topocount) [inline src/routes/routes.tsx:87]: Let's make this only navigable in a dev environment for now.

File: api/discord.ts
Comment (topocount) [inline api/discord.ts:20]: Can you provide some docstring comments as a high-level overview of what this does? A link to the documentation where this general flow is detailed would probably suffice to provide most of the lower level detail (Some thing like [this](https://discord.com/developers/docs/topics/oauth2#authorization-code-grant))

**Review by topocount — COMMENTED** [review]
Thanks @karelianpie this is coming together nicely. One request: Could you write out a detailed test plan? 

The test plan in the PR description isn't just a place to cite one's work, but ideally a place to detail the steps needed to verify that this set of changes does what it's supposed to (CLI commands, possibly starting with `yarn start`, in addition to configs or mocks that need to be run, and browser navigation). The intended audience would be a community developer who is curious to try out the PR but has little context as to how to get this started. This allows reviewers to dig into the details of your work with minimal overhead. I personally do this with my work when I know other developers will be looking at it closely before it gets merged.


[More details on test plans here](https://github.com/coordinape/coordinape/blob/main/CONTRIBUTING.md#test-plan)

File: src/routes/routes.tsx
Comment (levity) [reply src/routes/routes.tsx:87]: A feature flag would work too

File: hasura/migrations/default/1654202757203_alter_table_public_circle_api_keys_add_column_read_discord/down.sql
Comment (karelianpie) [reply hasura/migrations/default/1654202757203_alter_table_public_circle_api_keys_add_column_read_discord/down.sql]: Added a down migration with https://github.com/coordinape/coordinape/pull/1625/commits/c4056e70f74ae6a8678949429e777757c3e487ea

File: src/routes/routes.tsx
Comment (karelianpie) [reply src/routes/routes.tsx]: I like the idea of a feature flag, added the `discord` feature flag with https://github.com/coordinape/coordinape/pull/1625/commits/c41b0f7d6de2bebb96088df13aef67269b3983ca

**Comment by karelianpie** [comment]
> Thanks @karelianpie this is coming together nicely. One request: Could you write out a detailed test plan?
> 
> The test plan in the PR description isn't just a place to cite one's work, but ideally a place to detail the steps needed to verify that this set of changes does what it's supposed to (CLI commands, possibly starting with `yarn start`, in addition to configs or mocks that need to be run, and browser navigation). The intended audience would be a community developer who is curious to try out the PR but has little context as to how to get this started. This allows reviewers to dig into the details of your work with minimal overhead. I personally do this with my work when I know other developers will be looking at it closely before it gets merged.
> 
> [More details on test plans here](https://github.com/coordinape/coordinape/blob/main/CONTRIBUTING.md#test-plan)

Apologies, my bad, added the test plan now. Let me know if you run into any issues. I would suggest we should rewrite the prompt in the template, currently it's just

> How have you tested your changes? Are there additional steps required to deploy this change?

File: api/discord.ts
Comment (topocount) [inline api/discord.ts:45]: one last request before merging this. Please move these env vars to the [config file](https://github.com/karelianpie/coordinape/blob/e68b6a9b018036b13dfecb59409c16b43d3ca7e9/api-lib/config.ts) and parse them with fallbacks to minimize friction for devs. For the `DICORD_BOT_CLIENT*` fallbacks, just use something like `'no_token'` and for the `DISCORD_BOT_REDIRECT_URI` add the local redirect as instructed in the test plan.

**Review by topocount — DISMISSED** [review]
approved pending the last requested change below. ping me and I'll merge this.

File: api/discord.ts
Comment (karelianpie) [reply api/discord.ts]: Nice! Wasn't aware of this `config.ts`. Refactored with  https://github.com/coordinape/coordinape/pull/1625/commits/cd2de069a1c671dc52f1c84ba1bfc7d8c0425204

**Comment by levity** [comment]
@karelianpie in the future, consider splitting PRs like this into one that has schema/metadata changes, and one with frontend changes. did you try https://graphite.dev yet, btw?

**Comment by karelianpie** [comment]
> @karelianpie in the future, consider splitting PRs like this into one that has schema/metadata changes, and one with frontend changes. did you try https://graphite.dev yet, btw?

Will do, thanks for the heads up! I did try https://graphite.dev, but I think I might need to spend a bit more time on it as I didn't yet see much value in it. It's good to have an overview of all PRs and perform some actions like add reviewers/labels and so on a little bit faster but do you use it for something else @levity? I didn't find the resolving merging conflicts any easier on it for example

**Comment by levity** [comment]
@karelianpie the main use case for me is managing stacked branches. 

so e.g. with [split](https://docs.graphite.dev/guides/graphite-cli/squashing-folding-and-splitting#gt-branch-split) i can start with a normal branch but then split off some commits into a new branch, and the 2nd one is automatically connected to the first one

and if i add new commits to the "bottom" branch, there's a quick command to update the "top" branch, or as many branches as there are above it (`gt stack restack`)

and if there are new commits on main, i can run `gt repo sync --restack` to rebase the entire stack of branches, and then push them all up with `gt stack submit`

**Pull request merged**
